README
/*
 * README
 * mod_akismet.c - Spam Comments Detector Module for Apache2
 */

                        __          __   _                     __
   ____ ___  ____  ____/ /   ____ _/ /__(_)________ ___  ___  / /_
  / __ `__ \/ __ \/ __  /   / __ `/ //_/ / ___/ __ `__ \/ _ \/ __/
 / / / / / / /_/ / /_/ /   / /_/ / ,< / (__  ) / / / / /  __/ /_
/_/ /_/ /_/\____/\__,_/_______,_/_/|_/_/____/_/ /_/ /_/\___/\__/
                     /_____/

====================================================================
* Introduction
====================================================================

mod_akismet is Apache module that enables you to detect spam comments
before your application handler by Akismet anti-comment-spam service.
It is designed to stand in front of your Apache application handler
and allow you to check out spam comment post. If it detects spam post,
it adds spam detected flag to both apache subprocess_env and
incoming http header table with the key named "akismet_detect_spam".
Therefore, the spam detected flag can be referred by app handler or
next apache modules.

====================================================================
* Download
====================================================================

http://code.google.com/p/modakismet/downloads/list

====================================================================
* Prerequisites
====================================================================

o WordPress API Key - Akismet API requires an API key which are provided
  along with accounts to WordPress.com.
  http://wordpress.com/

  check out the url below for more information on the API key:
  http://akismet.com/development/api/

o libcurl - needed to compile mod_akismet
  http://curl.haxx.se/

====================================================================
* Build and Install
====================================================================

1) extract files from an archive
tar zxf modakismet-<VERSION>.tar.gz
cd modakismet-<VERSION>

2) open Makefile and modify ap_basedir variable
vi Makefile
 ap_basedir=/PATH-TO-APACHE-BASE

3) make and install
make
sudo make install

====================================================================
* Configuration Directives
====================================================================

The mod_akismet use shm and global mutex lock for caching the API Key
verfirication info, and directives on shm and lock must be specified
as per-server configuration.
However, all the other directives may be specified in anywhere like
Server, VirtualHost, Location, and so on.

Most of configurations are used for Akismet API params, therefore
better off checking out Akismet API documents as well.
http://akismet.com/development/api/


o AkismetApiCacheShmFile
    Description: [required] Set filename of shm file for the API Cache.
    Syntax: AkismetApiCacheShmFile filename
    Context: server config
    Status: Extension
    Module: mod_akismet

o AkismetGlobalLockFile
    Description: [required] Set filename of global mutex.
    Syntax: AkismetGlobalLockFile filename
    Context: server config
    Status: Extension
    Module: mod_akismet

o AkismetEnabled
    Description: [required] Set "On" to enable akismet engine, "Off" to disable.
    Syntax: AkismetEnabled On/Off
    Context: server config, virtual host, directory, .htaccess
    Status: Extension
    Module: mod_akismet

o AkismetApiKey
    Description: [required] Set the wordpress API key.
    Syntax: AkismetApiKey apiKey
    Context: server config, virtual host, directory, .htaccess
    Status: Extension
    Module: mod_akismet

o AkismetBlogURL
    Description:  [required] Set the front page or home URL of your blog or wiki, etc.
    Syntax: AkismetBlogURL blogURL
    Context: server config, virtual host, directory, .htaccess
    Status: Extension
    Module: mod_akismet

o AkismetCommentParamKey
    Description: [required] Set the param key name of the submitted comment.
    Syntax: AkismetCommentParamKey paramKeyName
    Context: server config, virtual host, directory, .htaccess
    Status: Extension
    Module: mod_akismet

o AkismetCommentAuthorParamKey
    Description: [optional] Set the param key name of the submitted name with the comment.
    Syntax: AkismetCommentAuthorParamKey paramKeyName
    Context: server config, virtual host, directory, .htaccess
    Status: Extension
    Module: mod_akismet

o AkismetCommentAuthorEmailParamKey
    Description: [optional] Set the param key name of submitted email address.
    Syntax: AkismetCommentAuthorEmailParamKey paramKeyName
    Context: server config, virtual host, directory, .htaccess
    Status: Extension
    Module: mod_akismet

o AkismetCommentAuthorURLParamKey
    Description: [optional] Set the param key name of the commenter URL.
    Syntax: AkismetCommentAuthorURLParamKey paramKeyName
    Context: server config, virtual host, directory, .htaccess
    Status: Extension
    Module: mod_akismet

o AkismetCommentPermalinkParamKey
    Description: [optional] Set the param key name of the permalink of the entry the comment was submitted to.
    Syntax: AkismetCommentPermalinkParamKey paramKeyName
    Context: server config, virtual host, directory, .htaccess
    Status: Extension
    Module: mod_akismet

====================================================================
* Sample Configuration
====================================================================

1) load module
LoadModule akismet_module modules/mod_akismet.so

2) add shm and lock configuration directives (per-server only!)
AkismetApiCacheShmFile   "/tmp/akismet_cache.shm"
AkismetGlobalLockFile    "/tmp/akismet.lock"

3) add Akismet API configuration directives
AkismetEnabled                    "On"
AkismetApiKey                     "fxcc21c088f2"
AkismetBlogURL                    "http://blog.yk55.com"
AkismetCommentParamKey            "comment"
AkismetCommentAuthorParamKey      "author"
AkismetCommentAuthorEmailParamKey "author_email"
AkismetCommentAuthorURLParamKey   "author_url"
AkismetCommentPermalinkParamKey   "permalink"

see also: conf/apache_akismet.conf

====================================================================
* More Configuration
====================================================================

There are some other configurations like this below. These are
things that seems no need to be dynamically configured in reloading,
therefore, statically defined.
However if need to be modified, do modify and recompile the module!

@ akismet.h
-------------------------
/* Akismet API's base domain to be used by mod_akismet */
#define AKISMET_API_SERVER                 "rest.akismet.com"

/* Akismet API's version to be used by mod_akismet */
#define AKISMET_API_VERSION                "1.1"

/* Akismet API's Port number */
#define AKISMET_API_PORT                   "80"

/* Akismet API's key verification method name */
#define AKISMET_API_METHOD_VERIFY_KEY      "verify-key"

/* Akismet API's spam comment check method name */
#define AKISMET_API_METHOD_COMMENT_CHECK   "comment-check"

/* The response string that Key-verification method returns if the key is valid */
#define AKISMET_API_RES_VERIFY_KEY_OK      "valid"

/* The response string that comment-check method returns if the comment is spam */
#define AKISMET_API_RES_COMMENT_IS_SPAM    "true"

/* HTTP request timeout in milsec */
#define AKISMET_API_REQ_TIMEOUT            (100)

/* UserAgent name to be used by mod_akismet in requesting to Akismet API */
#define AKISMET_API_REQ_USERAGENT          "mod_akismet/1.0"
-------------------------

====================================================================
* Application Sample
====================================================================

scripts/comment.php

====================================================================
* Authors
====================================================================
Yoichi Kawasaki <yokawasa@gmail.com>



Changelog
2009-11-28

    * fix: modules.mk

2009-5-10  [Version 0.0.1 released]

    * Test Release



akismet.c
/*
 * mod_akismet.c - Apache module that enables you to detect spam comments
 * before your application handler by Akismet anti-comment-spam service
 *
 * Copyright (C) 2009 Yoichi Kawasaki All rights reserved.
 * yk55.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <stdio.h>
#include <assert.h>
#include "akismet.h"
#include "http_log.h"      // ap_log_rerror
#include "apr_strings.h"
#include <curl/curl.h>

#define IS_EMPTY(a) ( (a != 0 && *a != 0) ? 0 : 1 )

struct akismet_curl_response{
    char *data;
    size_t size;
};

size_t
 akismet_write_response_callback(void *ptr,
        size_t size, size_t nmemb, void *data)
{
    register int rsize = 0;
    struct akismet_curl_response *res = NULL;
    rsize = size * nmemb;
    res = (struct akismet_curl_response *)data;
    res->data = (char *)realloc(res->data, res->size + rsize + 1);
    if (res->data) {
        memcpy(&(res->data[res->size]), ptr, rsize);
        res->size += rsize;
        res->data[res->size] = 0;
    }
    return rsize;
}

int
 akismet_http_post ( const char *url,
        const char *args, char* content, int* ret_code )
{
    CURL *ch;
    CURLcode retCode;
    // init struct
    struct akismet_curl_response res;
    res.data = NULL;
    res.size = 0;
    // curl init & set options
    curl_global_init(CURL_GLOBAL_ALL);
    ch = curl_easy_init();
    curl_easy_setopt(ch, CURLOPT_URL, url );
    curl_easy_setopt(ch, CURLOPT_TIMEOUT, AKISMET_API_REQ_TIMEOUT );
    curl_easy_setopt(ch, CURLOPT_POSTFIELDS, args );
    curl_easy_setopt(ch, CURLOPT_POST, 1);
    curl_easy_setopt(ch, CURLOPT_WRITEFUNCTION, akismet_write_response_callback );
    curl_easy_setopt(ch, CURLOPT_WRITEDATA, (void *)&res );
    curl_easy_setopt(ch, CURLOPT_USERAGENT, AKISMET_API_REQ_USERAGENT);
    // curl execute
    retCode = curl_easy_perform(ch);
    curl_easy_cleanup(ch);
    strncpy(content, res.data, res.size);
    *ret_code = retCode;
    free(res.data);

    if ( CURLE_OK != retCode || !content ) {
        return -1;
    }
    return 0;
}

char*
 url_encode( const char *src, char* dst )
{
    int s = 0;
    char *ch, *bk;
    char *p, *buf;
    static char escapechars[] = "<>{}#%|\"\\^~[]`@:\033";
    static char hex[16]  = "0123456789ABCDEF";
    bk = ch  = (char*)src;
    do{
        if( strchr( escapechars, *ch ))
            s += 2;
        ch++; s++;
    } while( *ch );

    buf = dst;
    p   = buf;
    ch  = bk;
    do{
        if( strchr( escapechars, *ch )){
            const char c = *ch;
            *p++ = '%';
            *p++ = hex[(c >> 4) & 0xf];
            *p++ = hex[c & 0xf];
        } else if(*ch == ' ') {
            *p++ = '+';
        } else{
            *p++ = *ch;
        }
        ch++;
    } while( *ch );
    *p = '\0';
    return( buf );
}

AkismetCode
 akismet_verify_key(request_rec *r,
            const char *apikey, const char* blogurl)
{
    int retcode;
    char api_res_content[100];
    char* apiurl;
    char* args;
    char blogurl_enc[200];

    if (!apikey||!blogurl) {
        return AKISMET_BAD_PARAM_ERROR;
    }
    memset(api_res_content,0,100);
    apiurl = apr_psprintf(r->pool, "http://%s/%s/%s",
                AKISMET_API_SERVER,AKISMET_API_VERSION,
                AKISMET_API_METHOD_VERIFY_KEY);
    args = apr_psprintf(r->pool, "key=%s&blog=%s",
                apikey, url_encode(blogurl, blogurl_enc));

    if (akismet_http_post(apiurl, args, api_res_content, &retcode ) ) {
        ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
                "akismet_verify_key FAILURE retcode=%d URL=%s ARGS=%s", retcode, apiurl, args);
        return AKISMET_HTTP_POST_ERROR;
    }
    if (strncmp(api_res_content,
            AKISMET_API_RES_VERIFY_KEY_OK,
            strlen(AKISMET_API_RES_VERIFY_KEY_OK)) != 0 ) {
        return AKISMET_KEY_INVALID;
    }
    return AKISMET_OK;
}

AkismetCode
 akismet_comment_check(request_rec *r, akismet_comment_check_request *acc_req )
{
    assert(acc_req);
    if (IS_EMPTY(acc_req->apikey)
        || IS_EMPTY(acc_req->blogurl)
        || IS_EMPTY(acc_req->comment)) {
        return AKISMET_BAD_PARAM_ERROR;
    }
    int retcode;
    char api_res_content[100];
    char* apiurl;
    char* args;
    size_t blogurl_enc_size= strlen(acc_req->blogurl) * 2 + 1;
    size_t ip_enc_size= (!IS_EMPTY(acc_req->ip)) ? strlen(acc_req->ip) * 2 + 1 : 1;
    size_t ua_enc_size= (!IS_EMPTY(acc_req->ua)) ? strlen(acc_req->ua) * 2 + 1 : 1;
    size_t ref_enc_size= (!IS_EMPTY(acc_req->ref)) ? strlen(acc_req->ref) * 2 + 1 : 1;
    char blogurl_enc[blogurl_enc_size];
    char ip_enc[ip_enc_size];
    char ua_enc[ua_enc_size];
    char ref_enc[ref_enc_size];

    memset(blogurl_enc, 0, blogurl_enc_size);
    memset(ip_enc, 0, ip_enc_size);
    memset(ua_enc, 0, ua_enc_size);
    memset(ref_enc, 0, ref_enc_size);
    memset(api_res_content,0,100);

    apiurl = apr_psprintf(r->pool, "http://%s.%s/%s/%s",
            acc_req->apikey,
            AKISMET_API_SERVER,
            AKISMET_API_VERSION,
            AKISMET_API_METHOD_COMMENT_CHECK);

    args = apr_psprintf( r->pool,
           "key=%s&blog=%s&user_ip=%s&user_agent=%s&referrer=%s"
           "&comment_content=%s&comment_author=%s&comment_author_email=%s"
           "&comment_author_url=%s&permalink=%s",
            acc_req->apikey,
            url_encode(acc_req->blogurl, blogurl_enc),
            !IS_EMPTY(acc_req->ip) ? url_encode(acc_req->ip, ip_enc) : "",
            !IS_EMPTY(acc_req->ua) ? url_encode(acc_req->ua, ua_enc) : "",
            !IS_EMPTY(acc_req->ref) ? url_encode(acc_req->ref, ua_enc) : "",
            acc_req->comment,
            !IS_EMPTY(acc_req->comment_author) ? acc_req->comment_author : "",
            !IS_EMPTY(acc_req->comment_author_email) ? acc_req->comment_author_email : "",
            !IS_EMPTY(acc_req->comment_author_url) ? acc_req->comment_author_url : "",
            !IS_EMPTY(acc_req->comment_permalink) ? acc_req->comment_permalink : "");
#ifdef AKISMET_DEBUG
ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r, "akismet_http_post ARGS->%s", args);
#endif
    if ( akismet_http_post(apiurl, args, api_res_content, &retcode ) ) {
        ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,  "FAILURE retcode->%d", retcode );
        return AKISMET_HTTP_POST_ERROR;
    }
    if ( strncmp( api_res_content,
            AKISMET_API_RES_COMMENT_IS_SPAM ,
            strlen(AKISMET_API_RES_COMMENT_IS_SPAM)
            ) == 0) {
        return AKISMET_SPAM;
    }
    return AKISMET_OK;
}

/*
 * vim:ts=4 et
 */



scripts/comment.php
<?php
$vars = $_POST;
$params=array(
    'author_name' => $vars['author_name'],
    'author_email' => $vars['author_email'],
    'author_url' => $vars['author_url'],
    'comment' => $vars['comment'],
    'done' => $vars['done'],
    'spam_detect' => getenv ( 'akismet_detect_spam' )
);
?>
<HTML>
<HEAD>
<TITLE>comment form sample</TITLE>
<META Http-Equiv="Content-Type" Content="text/html;charset=UTF-8">
</HEAD>
<BODY>
<H1 ALIGN="center">comment form</H1><P>
<FONT SIZE=5>please input comment and other information!</FONT><P>
<FORM METHOD="POST" ACTION="/comment.php">
<CENTER>
<TABLE BORDER WIDTH="90%">
    <TR>
        <TD NOWRAP>name
        </TD>
        <TD NOWRAP><INPUT TYPE="text" SIZE="40" NAME="author_name" VALUE="<?php echo $params['author_name']; ?>" >
        </TD>
    </TR>
    <TR>
        <TD NOWRAP>email
        </TD>
        <TD NOWRAP><INPUT TYPE="text" SIZE="50" NAME="author_email" VALUE="<?php echo $params['author_email']; ?>" >
        </TD>
    </TR>
    <TR>
        <TD NOWRAP>url
        </TD>
        <TD NOWRAP><INPUT TYPE="text" SIZE="40" NAME="author_url" VALUE="<?php echo $params['author_url']; ?>" >
        </TD>
    </TR>
    <TR>
        <TD VALIGN="top" NOWRAP>comment
        </TD>
        <TD NOWRAP><TEXTAREA NAME="comment" ROWS="10"COLS="35" WRAP="hard"><?php echo $params['comment']; ?></TEXTAREA>
        </TD>
    </TR>
</TABLE>
<P>
<INPUT type="submit" value="send">
<INPUT type="reset" name="clear" value="clear">
<INPUT TYPE="hidden" VALUE="1" NAME="done">


</CENTER>
</FORM>

<BR>
<PRE>
<?php
print_r($params);

if (!empty($params['done'])) {
    if ( !empty($params['spam_detect']) ) {
        echo "<p>SPAM!!!</p>";
    } else {
        echo "<p>OK!!!</p>";
    }
}
?>
</PRE>
</BODY>
</HTML>


akismet.h
/*
 * mod_akismet.c - Apache module that enables you to detect spam comments
 * before your application handler by Akismet anti-comment-spam service
 *
 * Copyright (C) 2009 Yoichi Kawasaki All rights reserved.
 * yk55.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#ifndef _AKISMET_H_
#define _AKISMET_H_

#include "httpd.h"

/* Akismet API's base domain to be used by mod_akismet */
#define AKISMET_API_SERVER                 "rest.akismet.com"
/* Akismet API's version to be used by mod_akismet */
#define AKISMET_API_VERSION                "1.1"
/* Akismet API's Port number */
#define AKISMET_API_PORT                   "80"
/* Akismet API's key verification method name */
#define AKISMET_API_METHOD_VERIFY_KEY      "verify-key"
/* Akismet API's spam comment check method name */
#define AKISMET_API_METHOD_COMMENT_CHECK   "comment-check"
/* The response string that Key-verification method returns if the key is valid */
#define AKISMET_API_RES_VERIFY_KEY_OK      "valid"
/* The response string that comment-check method returns if the comment is spam */
#define AKISMET_API_RES_COMMENT_IS_SPAM    "true"
/* HTTP request timeout in milsec */
#define AKISMET_API_REQ_TIMEOUT            (100)
/* UserAgent name to be used by mod_akismet in requesting to Akismet API */
#define AKISMET_API_REQ_USERAGENT          "mod_akismet/1.0"

typedef enum {
    AKISMET_OK   = 0,
    AKISMET_SPAM,
    AKISMET_KEY_INVALID,
    AKISMET_BAD_PARAM_ERROR,
    AKISMET_HTTP_POST_ERROR
}AkismetCode;

typedef struct {
    char *apikey;
    char *blogurl;
    char *ip;
    char *ua;
    char *ref;
    char *comment;
    char *comment_author;
    char *comment_author_email;
    char *comment_author_url;
    char *comment_permalink;
} akismet_comment_check_request;

AkismetCode
 akismet_verify_key(request_rec *r, const char *apikey, const char* blggurl);

AkismetCode
 akismet_comment_check(request_rec *r, akismet_comment_check_request *acc_req );


#endif
